### PR TITLE
SPARK-7021 Add JUnit output for Python unit tests

### PR DESCRIPTION
**Pull request by gliptak**
WIP

### CODE REVIEW

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

**Comment by gliptak** [comment]
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41290/console
Recording test results
ERROR: Publisher 'Publish JUnit test result report' failed: No test report files were found. Configuration error?
Test FAILed.

Can somebody comment on how is Jenkins configured to look for JUnit results (or did this build fail on RAT)? Thanks

**Comment by JoshRosen** [comment]
The build failed on RAT:

```
Could not find Apache license headers in the following files:
 !????? /home/jenkins/workspace/SparkPullRequestBuilder/dev/xmlrunner/__init__.py
 !????? /home/jenkins/workspace/SparkPullRequestBuilder/dev/xmlrunner/xmlrunner.py
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/dev/check-license ; received return code 1
```

**Comment by gliptak** [comment]
I just added the headers, xmlrunner is LGPL (and only used to run the tests).

**Comment by gliptak** [comment]
Is this a no-go because of the license? Thanks

**Comment by mengxr** [comment]
@gliptak Is https://pypi.python.org/pypi/unittest-xml-reporting/1.12.0 an alternative? It is BSD license. We can install the package on Jenkins machines, instead of including the source code in the codebase.

**Comment by gliptak** [comment]
@mengxr Would installing on Jenkins be independent from the codebase (the `import xmlrunner` just be available)? Thanks

**Comment by mengxr** [comment]
If this new library works, I think we could install it on Spark Jenkins. Could you test it locally and see whether it works?

**Comment by gliptak** [comment]
I tried this locally and the files were generated. Somebody with visibility on Jenkins might need to comment on file locations ...

**Comment by mengxr** [comment]
Please fix Python style issue:

~~~
./python/pyspark/streaming/tests.py:1303:13: E128 continuation line under-indented for visual indent
~~~

File: python/pyspark/streaming/tests.py [hunk 1]
```diff
@@ -1297,4 +1299,5 @@ def search_kinesis_asl_assembly_jar():
     for testcase in testcases:
         sys.stderr.write("[Running %s]\n" % (testcase))
         tests = unittest.TestLoader().loadTestsFromTestCase(testcase)
-        unittest.TextTestRunner(verbosity=3).run(tests)
+        unittest.main(tests, verbosity=3,
+            testRunner=xmlrunner.XMLTestRunner(output='target/surefire-reports'))
```
Comment (mengxr) [inline python/pyspark/streaming/tests.py]: this should be aligned with `tests,`. you can run `dev/lint-python` to check style locally.

**Comment by gliptak** [comment]
@mengxr Thanks for the local lint pointer

**Comment by mengxr** [comment]
@shaneknapp Could you take a look at this PR and see whether we can install https://pypi.python.org/pypi/unittest-xml-reporting/1.12.0 on Jenkins?

@gliptak We shouldn't make `xmlrunner` a requirement on local machines. If it is installed, we output XML. Otherwise, we should fallback to the default behavior. This could be done by catching exceptions at the import line.

**Comment by shaneknapp** [comment]
could you create a spark issue for the plugin and assign it to me?

also, i won't be able to do this until monday, as i'm about to leave town
for the weekend.

**Comment by gliptak** [comment]
@mengxr Revised.

**Comment by mengxr** [comment]
Created https://issues.apache.org/jira/browse/SPARK-10830.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
unittest-xml-reporting is installed on AMPLab Jenkins. So let's call Jenkins again.

**Comment by gliptak** [comment]
https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1810/testReport/

doesn't seem to list the Python tests.

Can somebody with access to Jenkins configuration review?

Thanks

**Comment by shaneknapp** [comment]
i'll take a look at the config tomorrow

**Comment by mengxr** [comment]
@gliptak We use `**/target/test-reports/*.xml` on Jenkins to locate XML reports. So `target/surefire-reports` may not work. (@shaneknapp please correct me if I'm wrong.)

**Comment by gliptak** [comment]
@mengxr I pushed a commit with the updated location.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
Though Jenkins failed, I can see the report now: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43107/testReport/junit/(root)/SerializationTestCase/test_pickling_file_handles/.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by gliptak** [comment]
@mengxr Thank you for guiding this

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
it seems that this PR has a higher-than-normal failure rate .. I don't know why. +@davies

**Comment by mengxr** [comment]
test this please

**Comment by JoshRosen** [comment]
Huh, looks like the last two failures were for the same PySpark test:

```
======================================================================
FAIL [0.000s]: test_pickling_file_handles (__main__.SerializationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/tests.py", line 259, in test_pickling_file_handles
    self.assertEqual(out1, out2)
AssertionError: <StringIO.StringIO instance at 0x2d65a28> != <StringIO.StringIO instance at 0x2ffdb90>
```

**Comment by JoshRosen** [comment]
I guess that this doesn't handle doctests, right? Is there any way to get XML for those, too? If not, could you edit the PR title / description to reflect this?

**Comment by JoshRosen** [comment]
@gliptak, can you try to reproduce the failure locally? See `./pyspark/run-tests --help` for hints on how to run only this test suite under Python 2.6.

**Comment by gliptak** [comment]
@JoshRosen I do not have python2.6 installed (or available as a package) locally

**Comment by gliptak** [comment]
@JoshRosen How are the doctest-s triggered on the commandline and in Jenkins?

```
$ python/run-tests --help
No handlers could be found for logger "root"
Usage: run-tests [options]

Options:
  -h, --help            show this help message and exit
  --python-executables=PYTHON_EXECUTABLES
                        A comma-separated list of Python executables to test
                        against (default: python,python3.4)
  --modules=MODULES     A comma-separated list of Python modules to test
                        (default: pyspark-core,pyspark-ml,pyspark-mllib
                        ,pyspark-sql,pyspark-streaming)
  -p PARALLELISM, --parallelism=PARALLELISM
                        The number of suites to test in parallel (default 4)
  --verbose             Enable additional debug logging
```

**Comment by gliptak** [comment]
What is the next step on this pull request? Thanks

**Comment by mengxr** [comment]
@gliptak I had an offline discussion with @davies . Let's do the following:

1. Change L259 in `pyspark/tests.py` to:

  ~~~python
    # Regression test for SPARK-3415
    def test_pickling_file_handles(self):
        if not xmlrunner:
            ser = CloudPickleSerializer()
            out1 = sys.stderr
            out2 = ser.loads(ser.dumps(out1))
            self.assertEqual(out1, out2)
~~~
Create a JIRA for this so we remember to fix it. The issue is `CloudPickeSerializer` wraps `stderr`, which conflicts with `xmlrunner`. But it might take some time to fix.
2. Change the PR title to `Python unit tests" since we do not report doctests.

Then we can run Jenkins again and see whether it works.

**Comment by davies** [comment]
Right now, all the Python unit tests are in the package of `(root)`, is there a way to set a package for it (such as pyspark)?

**Comment by gliptak** [comment]
@davies Not that I see. I opened https://github.com/xmlrunner/unittest-xml-reporting/issues/95 to discuss.

**Comment by gliptak** [comment]
@JoshRosen I opened https://github.com/xmlrunner/unittest-xml-reporting/issues/94 to discuss doctests processing.

**Comment by mengxr** [comment]
This looks good to me. I think we can address the test namespace issue in a later PR. @davies @JoshRosen ?

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

**Comment by gliptak** [comment]
As per https://github.com/xmlrunner/unittest-xml-reporting/issues/95 the Python tests do not have packages (or possibly not being recognized by https://github.com/xmlrunner/unittest-xml-reporting/

**Comment by gliptak** [comment]
Adding packages to XMLs at https://github.com/apache/spark/pull/9263

**Pull request closed without merging**